unit_test/main_function_test.py

```diff
@@ -0,0 +1,212 @@
+import sys
+import os
+import subprocess
+
+import pytest
+
+from main_function_test_utils import argtest, apply_mock_protection
+from cibuildwheel.__main__ import main
+from cibuildwheel import windows, linux, macos
+from cibuildwheel.util import BuildSelector
+from cibuildwheel.environment import ParsedEnvironment
+
+
+@pytest.mark.parametrize("system", ["macos", "linux", "windows"])
+@pytest.mark.parametrize("choose_platform_method", ["partameter", "environment"])
+def test_platform_chose(system, choose_platform_method, argtest, monkeypatch):
+    apply_mock_protection(monkeypatch)
+    monkeypatch.setattr(globals()[system], "build", argtest)
+    if choose_platform_method == "partameter":
+        monkeypatch.setattr(sys, "argv", sys.argv + ["--platform", system])
+    else:
+        monkeypatch.setattr(os, 'environ', {"CIBW_PLATFORM": system})
+    main()
+    assert argtest.kwargs["project_dir"] == "abcabc"
+
+@pytest.mark.parametrize("system", ["macos", "linux", "windows"])
+@pytest.mark.parametrize("output_dir", ["partameter", "environment", "both", "none"])
+def test_output_dir_set(system, output_dir, monkeypatch, argtest):
+    apply_mock_protection(monkeypatch)
+    monkeypatch.setattr(globals()[system], "build", argtest)
+    output_name = "out"
+    env = {"CIBW_PLATFORM": system}
+    if output_dir == "partameter":
+        monkeypatch.setattr(sys, "argv", sys.argv + ["--output-dir", output_name])
+    elif output_dir == "environment":
+        env["CIBW_OUTPUT_DIR"] = output_name
+    elif output_dir == "both":
+        monkeypatch.setattr(sys, "argv", sys.argv + ["--output-dir", output_name])
+        env["CIBW_OUTPUT_DIR"] = output_name + "aaa"
+    else:
+        output_name = "wheelhouse"
+    monkeypatch.setattr(os, 'environ', env)
+    main()
+    assert argtest.kwargs["output_dir"] == output_name
+
+@pytest.mark.parametrize("system", ["macos", "linux", "windows"])
+@pytest.mark.parametrize("build,build_set", [
+    ("*", {"aaaa", "abcd", "abab", "bcde"}),
+    ("a*", {"aaaa", "abcd", "abab"}),
+    ("*b*", {"abcd", "abab", "bcde"})
+    ])
+@pytest.mark.parametrize("skip,skip_set", [
+    ("", set()), ("a*", {"aaaa", "abcd", "abab"}),
+    ("*c*", {"abcd", "bcde"})
+])
+def test_build_selector(system, build, build_set, skip, skip_set, monkeypatch, argtest):
+    apply_mock_protection(monkeypatch)
+    monkeypatch.setattr(globals()[system], "build", argtest)
+    env = {"CIBW_PLATFORM": system, "CIBW_BUILD": build, "CIBW_SKIP": skip}
+    monkeypatch.setattr(os, 'environ', env)
+    main()
+    assert isinstance(argtest.kwargs["build_selector"], BuildSelector)
+    selector = argtest.kwargs["build_selector"]
+    selected = set([x for x in ["aaaa", "abcd", "abab", "bcde"] if selector(x)])
+    assert selected == build_set.difference(skip_set)
```

Comment:
We already have `build_selector_test.py` for this, for example, and integration tests to see if `CIBW_BUILD`/`CIBW_SKIP` work. Do we need this here again?

Reply:
I'm not sure. This check it faster, but of course there is redundancy.